File: frontend/src/services/raffle.ts
import api from './api';

export type RaffleAccessMode = 'guild_only' | 'world_only' | 'public';
export type RaffleScope = 'guild' | 'server' | 'global';
export type RaffleStatus = 'draft' | 'open' | 'closed' | 'completed' | 'cancelled' | 'deleted';

export interface RafflePrizeInput {
  name: string;
  reward: string;
  order_index?: number;
  position?: 'second' | 'first';
  amount?: number;
  currency?: string;
}

export interface RaffleParticipant {
  id: number;
  user_id?: number;
  guild_roster_character_id?: number;
  username?: string;
  character_name: string;
  normalized_character_name: string;
  account_identity_known: boolean;
  guild_rank?: string;
  weight: number;
  weight_multiplier: number;
  is_eligible: boolean;
  created_at: string;
  source?: string;
  eligibility_override?: boolean | null;
  eligibility_override_reason?: string;
}

export interface RafflePrize {
  id: number;
  name: string;
  reward: string;
  order_index: number;
  position?: 'second' | 'first';
  amount?: number;
  currency?: string;
}

export interface RaffleWinner {
  id: number;
  prize_id: number;
  prize_name: string;
  reward: string;
  participant_id: number;
  user_id: number;
  username: string;
  character_name: string;
  run_number: number;
  is_rerun: boolean;
  rerun_reason?: string;
  created_at: string;
}

export interface Raffle {
  id: number;
  public_code: string;
  title: string;
  description?: string;
  guild_name: string;
  scope_type: RaffleScope;
  world_name?: string;
  access_mode: RaffleAccessMode;
  show_participants: boolean;
  participant_count: number;
  visibility: 'public' | 'private';
  registration_enabled: boolean;
  run_mode: 'manual' | 'automatic';
  scheduled_run_at?: string;
  archive_after_days: number;
  archived_at?: string;
  status: RaffleStatus;
  current_run_number: number;
  rerun_count: number;
  created_at: string;
  updated_at: string;
  participants: RaffleParticipant[];
  prizes: RafflePrize[];
  current_winners: RaffleWinner[];
  history: RaffleWinner[];
  purpose: 'test' | 'real' | 'legacy';
  timezone_name: string;
  eligibility_days: number;
  publication_status: 'private' | 'published';
  execution_state: 'pending' | 'claimed' | 'running' | 'failed' | 'succeeded';
  executed_at?: string;
  scheduler_job_id?: string;
  claimed_at?: string;
  lease_expires_at?: string;
  last_error_code?: string;
  last_error_summary?: string;
  retry_count: number;
  next_retry_at?: string;
  unique_account_participation: boolean;
  weighting_mode: 'equal' | 'weighted';
}

export interface RaffleCandidate {
  roster_character_id: number; character_name: string; rank?: string; level?: number;
  vocation?: string; last_activity_at: string; linked_user_id?: number; linked_username?: string;
  account_identity_key?: string; account_identity_known: boolean; already_participating: boolean;
  selectable: boolean; reason?: string;
}

export interface EligibilityPreview {
  raffle_id: number; cutoff_at: string; timezone_name: string; eligibility_days: number;
  candidate_count: number; eligible_count: number; excluded_count: number; snapshot_hash: string;
  persisted: boolean; snapshot_id?: number;
  entries: Array<{ character_name?: string; guild_name?: string; guild_rank?: string; last_activity_at?: string; is_eligible: boolean; exclusion_code?: string; exclusion_summary?: string }>;
}

export interface AutomaticResult {
  id: number; prize_id: number; prize_position: 'second' | 'first'; prize_name: string;
  amount: number; currency: string; character_name: string; selection_index: number;
  candidate_count: number; delivery_status: 'pending' | 'delivered' | 'disputed' | 'cancelled'; delivery_deadline_at: string;
  delivered_at?: string; delivered_by_name?: string; delivery_note?: string;
  delivery_history?: Array<{ previous_status: string; new_status: string; actor: string; note?: string; admin_override: boolean; created_at: string }>;
}

export interface AutomaticRun {
  id: number; raffle_id: number; run_number: number; snapshot_id: number; parent_run_id?: number;
  trigger: string; state: string; started_at?: string; completed_at?: string; failure_code?: string;
  failure_summary?: string; algorithm_version: string; entropy_commitment?: string; results: AutomaticResult[];
}

export interface RaffleWorkspaceItem {
  id: number; public_code: string; title: string; guild_name: string; scope_type: RaffleScope; world_name?: string;
  purpose: 'test' | 'real' | 'legacy'; status: RaffleStatus; scheduled_run_at?: string;
  publication_status: 'private' | 'published'; execution_state: Raffle['execution_state']; participant_count: number;
  last_error_summary?: string; retry_count: number;
  eligibility?: { candidate_count: number; eligible_count: number; excluded_count: number; cutoff_at: string; frozen: boolean };
  winners: Array<{ prize_position: 'second' | 'first'; prize_name: string; amount: number; currency: string; character_name: string; delivery_status: AutomaticResult['delivery_status']; delivery_deadline_at: string }>;
  capabilities: { manage: boolean; publish: boolean };
  actions?: {
    view_detail: { enabled: boolean; reason?: string | null };
    edit: { enabled: boolean; reason?: string | null };
    edit_prizes: { enabled: boolean; reason?: string | null };
    cancel_archive: { enabled: boolean; reason?: string | null };
    permanent_delete: { enabled: boolean; reason?: string | null };
  };
}

export interface RaffleExecution {
  raffle_id: number;
  run_number: number;
  winner_count: number;
  winners: RaffleWinner[];
}

export interface RaffleSimulation {
  raffle_id: number;
  run_number: number;
  winner_count: number;
  winners: RaffleWinner[];
  simulation: true;
  status: RaffleStatus;
  access_mode: RaffleAccessMode;
  eligible_count: number;
  ineligible_count: number;
  participant_count: number;
  prizes: Array<{ id: number; name: string; reward: string; order_index: number }>;
  eligible_participants: Array<{
    participant_id: number;
    user_id: number;
    username: string;
    character_name: string;
    guild_rank?: string;
    weight: number;
    weight_multiplier: number;
  }>;
  ineligible_participants: Array<{ participant_id: number; user_id?: number; character_name: string; guild_rank?: string; weight: number; weight_multiplier: number; reason: string }>;
  warnings: string[];
}

export interface PublicRaffle {
  public_code: string;
  title: string;
  description?: string;
  guild_name: string;
  access_mode: RaffleAccessMode;
  purpose: 'test' | 'real' | 'legacy';
  timezone_name: string;
  scheduled_run_at?: string;
  status: RaffleStatus;
  publication_status: 'private' | 'reviewed' | 'published';
  show_participants: boolean;
  participant_count: number;
  participants: Array<{ character_name: string; guild_rank?: string }>;
  prizes: RafflePrize[];
  winners: Array<{
    prize_position: 'second' | 'first';
    prize_name: string;
    amount: number;
    currency: string;
    character_name: string;
    delivery_status: 'pending' | 'delivered' | 'disputed' | 'cancelled';
    delivery_deadline_at: string;
  }>;
}

export const raffleApi = {
  async workspace(): Promise<RaffleWorkspaceItem[]> { return (await api.get('/raffles/workspace')).data; },
  async list(): Promise<Raffle[]> {
    const response = await api.get('/raffles/');
    return response.data;
  },

  async get(raffleId: number): Promise<Raffle> {
    const response = await api.get(`/raffles/${raffleId}`);
    return response.data;
  },

  async update(raffleId: number, payload: Partial<{ title: string; description: string; guild_name: string; access_mode: RaffleAccessMode; show_participants: boolean; visibility: string; registration_enabled: boolean; run_mode: string; scheduled_run_at: string; timezone_name: string; eligibility_days: number; archive_after_days: number; status: RaffleStatus; unique_account_participation: boolean; weighting_mode: 'equal' | 'weighted' }>): Promise<Raffle> {
    const response = await api.put(`/raffles/${raffleId}`, payload);
    return response.data;
  },

  async share(raffleId: number): Promise<{ public_code: string; url: string }> {
    const response = await api.get(`/raffles/${raffleId}/share`);
    return response.data;
  },

  async create(payload: { title: string; description?: string; guild_name: string; scope_type?: RaffleScope; world_name?: string; access_mode?: RaffleAccessMode; show_participants?: boolean; prizes: RafflePrizeInput[]; purpose?: 'test' | 'real' | 'legacy'; run_mode?: 'manual' | 'automatic'; scheduled_run_at?: string; timezone_name?: string; eligibility_days?: number; unique_account_participation?: boolean; weighting_mode?: 'equal' | 'weighted' }): Promise<Raffle> {
    const response = await api.post('/raffles/', payload);
    return response.data;
  },

  async addPrize(raffleId: number, payload: RafflePrizeInput): Promise<Raffle> {
    const response = await api.post(`/raffles/${raffleId}/prizes`, payload);
    return response.data;
  },

  async syncParticipants(raffleId: number): Promise<Raffle> {
    const response = await api.post(`/raffles/${raffleId}/participants/sync`);
    return response.data;
  },

  async addManualParticipant(raffleId: number, characterName: string): Promise<Raffle> {
    const response = await api.post(`/raffles/${raffleId}/participants/manual`, { character_name: characterName });
    return response.data;
  },

  async updateWeight(raffleId: number, participantId: number, weight: number): Promise<Raffle> {
    const response = await api.patch(`/raffles/${raffleId}/participants/${participantId}/weight`, { weight });
    return response.data;
  },

  async candidates(raffleId: number, days: 7 | 15 | 30, search?: string): Promise<RaffleCandidate[]> {
    return (await api.get(`/raffles/${raffleId}/participant-candidates`, { params: { days, search } })).data;
  },
  async refreshGuildRoster(raffleId: number): Promise<Record<string, number | string>> {
    return (await api.post(`/raffles/${raffleId}/guild-roster/sync`)).data;
  },
  async addRosterParticipants(raffleId: number, rosterCharacterIds: number[], replaceExisting = false, addAllEligible = false, activityDays: 7 | 15 | 30 = 30): Promise<{ added: number; restored: number; removed: number; unchanged: number }> {
    return (await api.post(`/raffles/${raffleId}/participants/bulk`, { roster_character_ids: rosterCharacterIds, replace_existing: replaceExisting, add_all_eligible: addAllEligible, activity_days: activityDays })).data;
  },
  async removeParticipants(raffleId: number, participantIds: number[], reason?: string): Promise<{ removed: number }> {
    return (await api.post(`/raffles/${raffleId}/participants/remove`, { participant_ids: participantIds, reason })).data;
  },
  async updateParticipationSettings(raffleId: number, payload: { unique_account_participation?: boolean; weighting_mode?: 'equal' | 'weighted' }): Promise<Raffle> {
    return (await api.patch(`/raffles/${raffleId}/participation-settings`, payload)).data;
  },
  async manageableGuildContext(): Promise<{ guilds: string[]; guild_worlds: Record<string, string | null> }> {
    return (await api.get('/guild-management/manageable-guilds', { params: { capability: 'raffles.manage' } })).data;
  },
  async manageableGuilds(): Promise<string[]> {
    return (await api.get('/guild-management/manageable-guilds', { params: { capability: 'raffles.manage' } })).data.guilds;
  },

  async draw(raffleId: number, dryRun: boolean = false): Promise<RaffleExecution> {
    const response = await api.post(`/raffles/${raffleId}/draw`, { dry_run: dryRun });
    return response.data;
  },

  async rerun(raffleId: number, reason: string): Promise<RaffleExecution> {
    const response = await api.post(`/raffles/${raffleId}/rerun`, { reason });
    return response.data;
  },

  async removeParticipant(raffleId: number, participantId: number): Promise<Raffle> {
    const response = await api.delete(`/raffles/${raffleId}/participants/${participantId}`);
    return response.data;
  },

  async getPublic(raffleId: number): Promise<PublicRaffle> {
    const response = await api.get(`/raffles/public/${raffleId}`);
    return response.data;
  },

  async getPublicByCode(publicCode: string): Promise<PublicRaffle> {
    const response = await api.get(`/raffles/public/code/${publicCode}`);
    return response.data;
  },

  async registerPublic(raffleId: number, characterName: string): Promise<PublicRaffle> {
    const response = await api.post(`/raffles/public/${raffleId}/register`, { character_name: characterName });
    return response.data;
  },

  async registerPublicByCode(publicCode: string, characterName: string): Promise<PublicRaffle> {
    const response = await api.post(`/raffles/public/code/${publicCode}/register`, { character_name: characterName });
    return response.data;
  },

  async simulate(raffleId: number): Promise<RaffleSimulation> {
    const response = await api.post(`/raffles/${raffleId}/simulate`);
    return response.data;
  },

  async softDelete(raffleId: number, reason?: string): Promise<Raffle> {
    const response = await api.delete(`/raffles/${raffleId}`, { params: { reason } });
    return response.data;
  },

  async permanentDelete(raffleId: number, reason: string, confirmation: string): Promise<{ id: number; deleted: true; audit_preserved: boolean }> {
    const response = await api.delete(`/raffles/${raffleId}/permanent`, { params: { reason, confirmation } });
    return response.data;
  },
  async previewEligibility(raffleId: number): Promise<EligibilityPreview> {
    const response = await api.post(`/raffles/${raffleId}/eligibility/preview`);
    return response.data;
  },
  async freezeEligibility(raffleId: number): Promise<EligibilityPreview> {
    const response = await api.post(`/raffles/${raffleId}/eligibility/freeze`);
    return response.data;
  },
  async runs(raffleId: number): Promise<AutomaticRun[]> {
    const response = await api.get(`/raffles/${raffleId}/runs`);
    return response.data;
  },
  async rerunAutomatic(raffleId: number, positions: Array<'second' | 'first'>, reason: string, overrideDelivered = false, overrideReason?: string): Promise<AutomaticRun> {
    const response = await api.post(`/raffles/${raffleId}/reruns`, { positions, reason, override_delivered: overrideDelivered, override_reason: overrideReason });
    return response.data;
  },
  async publish(raffleId: number): Promise<void> { await api.post(`/raffles/${raffleId}/publish`); },
  async unpublish(raffleId: number): Promise<void> { await api.post(`/raffles/${raffleId}/unpublish`); },
  async updateDelivery(raffleId: number, resultId: number, status: AutomaticResult['delivery_status'], note?: string, adminOverride = false): Promise<void> {
    await api.patch(`/raffles/${raffleId}/results/${resultId}/delivery`, { status, note, admin_override: adminOverride });
  },
  async overrideTestEligibility(raffleId: number, participantId: number, eligible: boolean, reason: string): Promise<void> {
    await api.patch(`/raffles/${raffleId}/participants/${participantId}/test-eligibility-override`, { eligible, reason });
  },
  async retryTest(raffleId: number, reason: string): Promise<void> {
    await api.post(`/raffles/${raffleId}/test-retry`, { reason });
  },
  async cleanupTest(raffleId: number, reason: string): Promise<{ raffle_id: number; archived: boolean; participant_associations_removed: number; users_modified: number; guilds_modified: number; real_raffles_modified: number }> {
    return (await api.post(`/raffles/${raffleId}/test-cleanup`, { confirmation: 'ARCHIVE TEST RAFFLE', reason })).data;
  },
};

File: backend/app/schemas/raffle.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


class RafflePrizeCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    reward: str = Field(..., min_length=1, max_length=500)
    order_index: Optional[int] = None
    position: Optional[Literal["second", "first"]] = None
    amount: Optional[Decimal] = Field(None, gt=0)
    currency: Optional[str] = Field(None, min_length=1, max_length=20)


class RaffleCreate(BaseModel):
    title: str = Field(..., min_length=3, max_length=200)
    description: Optional[str] = None
    guild_name: str = Field(..., min_length=1)
    scope_type: Literal["guild", "server", "global"] = "guild"
    world_name: Optional[str] = None
    access_mode: Literal["guild_only", "world_only", "public"] = "guild_only"
    show_participants: bool = True
    visibility: Literal["public", "private"] = "public"
    registration_enabled: bool = True
    run_mode: Literal["manual", "automatic"] = "manual"
    scheduled_run_at: Optional[datetime] = None
    archive_after_days: int = 7
    prizes: List[RafflePrizeCreate] = Field(default_factory=list, max_length=50)
    purpose: Literal["test", "real", "legacy"] = "legacy"
    timezone_name: str = "America/Chicago"
    eligibility_days: int = Field(5, ge=1, le=30)
    eligibility_cutoff_at: Optional[datetime] = None
    unique_account_participation: bool = True
    weighting_mode: Literal["equal", "weighted"] = "equal"

    @field_validator("scheduled_run_at", "eligibility_cutoff_at")
    @classmethod
    def require_aware_timestamp(cls, value: Optional[datetime]) -> Optional[datetime]:
        if value is not None and (value.tzinfo is None or value.utcoffset() is None):
            raise ValueError("Timestamp must include a UTC offset")
        return value


class RaffleUpdate(BaseModel):
    expected_version: Optional[int] = Field(None, ge=1)
    title: Optional[str] = Field(None, min_length=3, max_length=200)
    description: Optional[str] = None
    guild_name: Optional[str] = None
    access_mode: Optional[str] = None
    show_participants: Optional[bool] = None
    visibility: Optional[str] = None
    registration_enabled: Optional[bool] = None
    run_mode: Optional[str] = None
    scheduled_run_at: Optional[datetime] = None
    archive_after_days: Optional[int] = None
    status: Optional[str] = None
    timezone_name: Optional[str] = None
    eligibility_days: Optional[int] = Field(None, ge=1, le=30)
    publication_status: Optional[Literal["private", "published"]] = None
    unique_account_participation: Optional[bool] = None
    weighting_mode: Optional[Literal["equal", "weighted"]] = None
    prizes: Optional[List[RafflePrizeCreate]] = Field(None, max_length=50)

    @field_validator("scheduled_run_at")
    @classmethod
    def require_aware_timestamp(cls, value: Optional[datetime]) -> Optional[datetime]:
        if value is not None and (value.tzinfo is None or value.utcoffset() is None):
            raise ValueError("Timestamp must include a UTC offset")
        return value


class RaffleParticipantResponse(BaseModel):
    id: int
    user_id: Optional[int] = None
    guild_roster_character_id: Optional[int] = None
    username: Optional[str] = None
    character_name: str
    normalized_character_name: str
    account_identity_known: bool = False
    guild_rank: Optional[str] = None
    weight: Decimal
    weight_multiplier: Decimal = Decimal("1")
    effective_probability: Optional[Decimal] = None
    is_eligible: bool
    created_at: datetime
    source: Optional[str] = None
    eligibility_override: Optional[bool] = None
    eligibility_override_reason: Optional[str] = None

    class Config:
        from_attributes = True


class RafflePrizeResponse(BaseModel):
    id: int
    name: str
    reward: str
    order_index: int
    position: Optional[str] = None
    amount: Optional[Decimal] = None
    currency: Optional[str] = None

    class Config:
        from_attributes = True


class RaffleWinnerResponse(BaseModel):
    id: Optional[int] = None
    prize_id: int
    prize_name: str
    reward: str
    participant_id: int
    user_id: Optional[int] = None
    username: Optional[str] = None
    character_name: str
    run_number: int
    is_rerun: bool
    rerun_reason: Optional[str] = None
    created_at: Optional[datetime] = None


class RaffleResponse(BaseModel):
    id: int
    public_code: str
    title: str
    description: Optional[str] = None
    guild_name: str
    scope_type: str = "guild"
    world_name: Optional[str] = None
    access_mode: str
    show_participants: bool
    participant_count: int
    visibility: str
    registration_enabled: bool
    run_mode: str
    scheduled_run_at: Optional[datetime] = None
    archive_after_days: int
    archived_at: Optional[datetime] = None
    status: str
    current_run_number: int
    rerun_count: int
    created_at: datetime
    updated_at: datetime
    participants: List[RaffleParticipantResponse] = Field(default_factory=list)
    prizes: List[RafflePrizeResponse] = Field(default_factory=list)
    current_winners: List[RaffleWinnerResponse] = Field(default_factory=list)
    history: List[RaffleWinnerResponse] = Field(default_factory=list)
    purpose: str = "legacy"
    timezone_name: str = "America/Chicago"
    eligibility_days: int = 5
    eligibility_cutoff_at: Optional[datetime] = None
    publication_status: str = "private"
    execution_state: str = "pending"
    executed_at: Optional[datetime] = None
    scheduler_job_id: Optional[str] = None
    claimed_at: Optional[datetime] = None
    lease_expires_at: Optional[datetime] = None
    last_error_code: Optional[str] = None
    last_error_summary: Optional[str] = None
    retry_count: int = 0
    next_retry_at: Optional[datetime] = None
    unique_account_participation: bool = True
    weighting_mode: Literal["equal", "weighted"] = "equal"
    version: int = 1


class RaffleExecutionResponse(BaseModel):
    raffle_id: int
    run_number: int
    winner_count: int
    winners: List[RaffleWinnerResponse]
    simulation: bool = False
    status: Optional[str] = None
    access_mode: Optional[str] = None
    eligible_count: Optional[int] = None
    ineligible_count: Optional[int] = None
    participant_count: Optional[int] = None
    prizes: Optional[List[dict]] = None
    eligible_participants: Optional[List[dict]] = None
    ineligible_participants: Optional[List[dict]] = None
    warnings: Optional[List[str]] = None


class RaffleRerunRequest(BaseModel):
    reason: str = Field(..., min_length=3)


class RaffleWeightUpdateRequest(BaseModel):
    weight: Decimal = Field(..., gt=0, le=1_000_000)


class RaffleParticipationSettingsRequest(BaseModel):
    unique_account_participation: Optional[bool] = None
    weighting_mode: Optional[Literal["equal", "weighted"]] = None


class RaffleParticipantsMutationRequest(BaseModel):
    roster_character_ids: List[int] = Field(default_factory=list, max_length=500)
    add_all_eligible: bool = False
    activity_days: Literal[7, 15, 30] = 30
    replace_existing: bool = False


class RaffleParticipantsRemoveRequest(BaseModel):
    participant_ids: List[int] = Field(..., min_length=1, max_length=500)
    reason: Optional[str] = Field(None, max_length=500)


class RaffleParticipantMutationResponse(BaseModel):
    raffle_id: int
    added: int = 0
    restored: int = 0
    removed: int = 0
    unchanged: int = 0


class RaffleCandidateResponse(BaseModel):
    roster_character_id: int
    character_name: str
    rank: Optional[str] = None
    level: Optional[int] = None
    vocation: Optional[str] = None
    last_activity_at: datetime
    linked_user_id: Optional[int] = None
    linked_username: Optional[str] = None
    account_identity_key: Optional[str] = None
    account_identity_known: bool
    already_participating: bool
    selectable: bool
    reason: Optional[str] = None


class RaffleDrawRequest(BaseModel):
    dry_run: bool = False


class EligibilityEntryResponse(BaseModel):
    character_name: Optional[str] = None
    guild_name: Optional[str] = None
    guild_rank: Optional[str] = None
    last_activity_at: Optional[datetime] = None
    is_eligible: bool
    exclusion_code: Optional[str] = None
    exclusion_summary: Optional[str] = None


class EligibilityPreviewResponse(BaseModel):
    raffle_id: int
    cutoff_at: datetime
    timezone_name: str
    eligibility_days: int
    candidate_count: int
    eligible_count: int
    excluded_count: int
    snapshot_hash: str
    entries: List[EligibilityEntryResponse]
    persisted: bool = False
    snapshot_id: Optional[int] = None


class AutomaticExecutionRequest(BaseModel):
    # The public management API is always a manual trigger. The future
    # scheduler will call the same service directly with trigger="scheduler".
    trigger: Literal["manual"] = "manual"


class AutomaticResultResponse(BaseModel):
    id: int
    prize_id: int
    prize_position: str
    prize_name: str
    amount: Decimal
    currency: str
    character_name: str
    selection_index: int
    candidate_count: int
    delivery_status: str
    delivery_deadline_at: datetime
    delivered_at: Optional[datetime] = None
    delivered_by_name: Optional[str] = None
    delivery_note: Optional[str] = None
    delivery_history: List[dict] = Field(default_factory=list)


class AutomaticRunResponse(BaseModel):
    id: int
    raffle_id: int
    run_number: int
    snapshot_id: int
    parent_run_id: Optional[int] = None
    trigger: str
    state: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    failure_code: Optional[str] = None
    failure_summary: Optional[str] = None
    algorithm_version: str
    entropy_commitment: Optional[str] = None
    results: List[AutomaticResultResponse] = Field(default_factory=list)


class AutomaticRerunRequest(BaseModel):
    positions: List[Literal["second", "first"]] = Field(..., min_length=1, max_length=2)
    reason: str = Field(..., min_length=3, max_length=1000)
    override_delivered: bool = False
    override_reason: Optional[str] = Field(None, max_length=1000)


class DeliveryUpdateRequest(BaseModel):
    status: Literal["pending", "delivered", "disputed", "cancelled"]
    note: Optional[str] = Field(None, max_length=2000)
    admin_override: bool = False


class DeliveryResponse(BaseModel):
    result_id: int
    status: str
    delivery_deadline_at: datetime
    delivered_at: Optional[datetime] = None
    delivered_by_id: Optional[int] = None
    note: Optional[str] = None


class TestEligibilityOverrideRequest(BaseModel):
    eligible: bool
    reason: str = Field(..., min_length=3, max_length=1000)


class TestRetryRequest(BaseModel):
    reason: str = Field(..., min_length=3, max_length=1000)


class TestCleanupRequest(BaseModel):
    confirmation: Literal["ARCHIVE TEST RAFFLE"]
    reason: str = Field(..., min_length=3, max_length=1000)


class TestCleanupResponse(BaseModel):
    raffle_id: int
    archived: bool
    participant_associations_removed: int
    users_modified: int = 0
    guilds_modified: int = 0
    real_raffles_modified: int = 0


class ManagerGrantRequest(BaseModel):
    user_id: int


class ManagerGrantResponse(BaseModel):
    raffle_id: int
    user_id: int
    granted_by_id: int
    created_at: datetime
    revoked_at: Optional[datetime] = None


class PublicationResponse(BaseModel):
    raffle_id: int
    publication_status: str
    published_at: Optional[datetime] = None
    published_by_id: Optional[int] = None


class PublicRaffleParticipantResponse(BaseModel):
    character_name: str
    guild_rank: Optional[str] = None


class PublicRaffleWinnerResponse(BaseModel):
    prize_position: str
    prize_name: str
    amount: Decimal
    currency: str
    character_name: str
    delivery_status: str
    delivery_deadline_at: datetime


class PublicRaffleResponse(BaseModel):
    public_code: str
    title: str
    description: Optional[str] = None
    guild_name: str
    access_mode: str
    purpose: str
    timezone_name: str
    scheduled_run_at: Optional[datetime] = None
    status: str
    publication_status: str
    show_participants: bool
    participant_count: int
    participants: List[PublicRaffleParticipantResponse] = Field(default_factory=list)
    prizes: List[RafflePrizeResponse] = Field(default_factory=list)
    winners: List[PublicRaffleWinnerResponse] = Field(default_factory=list)
